gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/webmaster/GoogleWebmasterDataFetcherImpl.java

```diff
@@ -0,0 +1,366 @@
+      LOG.info(String.format("A total of %d pages fetched for property %s at country-%s from %s to %s", actualSize,
+          _siteProperty, country, startDate, endDate));
+    } else {
+      int expectedSize = getPagesSize(startDate, endDate, country, requestedDimensions, Arrays.asList(countryFilter));
+      LOG.info(String.format("Total number of pages is %d for market-%s from %s to %s", expectedSize,
+          GoogleWebmasterFilter.countryFilterToString(countryFilter), startDate, endDate));
+      Queue<Pair<String, FilterOperator>> jobs = new ArrayDeque<>();
+      expandJobs(jobs, _siteProperty);
+
+      allPages = getPages(startDate, endDate, requestedDimensions, countryFilter, jobs);
+      allPages.add(_siteProperty);
+      actualSize = allPages.size();
+      if (actualSize != expectedSize) {
+        LOG.warn(
+            String.format("Expected page size for country-%s is %d, but only able to get %d", country, expectedSize,
+                actualSize));
+      }
+      LOG.info(String.format("A total of %d pages fetched for property %s at country-%s from %s to %s", actualSize,
+          _siteProperty, country, startDate, endDate));
+    }
+
+    ArrayDeque<ProducerJob> jobs = new ArrayDeque<>(actualSize);
+    for (String page : allPages) {
+      jobs.add(new SimpleProducerJob(page, startDate, endDate));
+    }
+    return jobs;
+  }
+
+  private int getPagesSize(final String startDate, final String endDate, final String country,
+      final List<Dimension> requestedDimensions, final List<ApiDimensionFilter> apiDimensionFilters)
+      throws IOException {
+
+    int startRow = 0;
+    int count = 4;
+    final ExecutorService es = Executors.newFixedThreadPool(count);
+    int r = 0;
+    while (r < 50) {
+      r++;
+      List<Future<Integer>> results = new ArrayList<>(count);
+
+      for (int i = 0; i < count; ++i) {
+        startRow += GoogleWebmasterClient.API_ROW_LIMIT;
+        final int start = startRow;
+        Future<Integer> submit = es.submit(new Callable<Integer>() {
+          @Override
+          public Integer call() {
+            LOG.info(String.format("Getting page size from %s...", start));
+            while (true) {
+              try {
+                List<String> pages =
+                    _client.getPages(_siteProperty, startDate, endDate, country, GoogleWebmasterClient.API_ROW_LIMIT,
+                        requestedDimensions, apiDimensionFilters, start);
+                if (pages.size() < GoogleWebmasterClient.API_ROW_LIMIT) {
+                  //Figured out the size
+                  return pages.size() + start;
+                } else {
+                  //The size is not determined.
+                  return -1;
+                }
+              } catch (IOException e) {
+                LOG.info(String.format("Getting page size from %s failed. Retrying...", start));
+              }
+              try {
+                Thread.sleep(200);
```

Comment:
No. This will not retry forever. Check this line **Integer integer = result.get(2, TimeUnit.MINUTES);**. The job will timeout in two minutes. But I will improve the thread exit based on interruption. And in this case, I don't need exponential back-off. Linear is good enough for me. Thank you for pointing that out, I will utilize that library in the future when I need exponential back-off.